tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc

```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+      LOG(ERROR) << "Allocating output failed with " << status;
+      ctx->SetStatus(status);
+      return;
     }
-
-    OP_REQUIRES_OK(context,
-                   context->allocate_output(i, output_shape, &output_tensor));
     auto dtype = trt_engine_ptr_->getBindingDataType(binding_index);
     switch (dtype) {
       case nvinfer1::DataType::kFLOAT:
         buffers[binding_index] =
             reinterpret_cast<void*>(output_tensor->flat<float>().data());
         break;
       case nvinfer1::DataType::kHALF:
-        LOG(FATAL) << "half size is not supported yet!";
+        LOG(ERROR) << "half size is not supported yet!";
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "Half outputs are not supported!"));
+        return;
         break;
       case nvinfer1::DataType::kINT8:
-        LOG(FATAL) << "int8 is not supported yet!";
+        LOG(ERROR) << "int8 is not supported yet!";
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "INT8 outputs are not supported!"));
+        return;
         break;
       default:
-        LOG(FATAL) << "Unknown data type: " << int(dtype);
+        LOG(ERROR) << "Unknown TRT data type: " << int(dtype);
+        ctx->SetStatus(tensorflow::errors::InvalidArgument(
+            "Unsupported output data type! " + int(dtype)));
+        return;
         break;
     }
   }
   // copied from cuda_kernel_helper since it seems only valid in *.cu.cc files
   const cudaStream_t* stream = CHECK_NOTNULL(
-      reinterpret_cast<const cudaStream_t*>(context->op_device_context()
+      reinterpret_cast<const cudaStream_t*>(ctx->op_device_context()
                                                 ->stream()
                                                 ->implementation()
                                                 ->CudaStreamMemberHack()));
 
   // TODO(jie): trt enqueue does not return error
-  auto ret = trt_execution_context_ptr_->enqueue(num_batch, &buffers[0],
-                                                 *stream, nullptr);
+  auto trt_execution_context_ptr = engine_ctx_pair.second;
+  auto ret = trt_execution_context_ptr->enqueue(num_batch, &buffers[0], *stream,
+                                                nullptr);
   VLOG(2) << "enqueue returns: " << ret;
   // sync should be done by TF.
-}
+}  // namespace tensorrt
 TRTEngineOp::~TRTEngineOp() {
   // Order matters!
-  trt_execution_context_ptr_.reset();
-  trt_engine_ptr_.reset();
-  allocator_.reset();
+  for (auto eng : engine_map) {
+    eng.second.first.reset();
+    eng.second.second.reset();
+  }
+  for (auto alloc : allocators_) alloc.second.reset();
+}
+// template <typename T>
+// using destroyed_ptr = std::shared_ptr<T, TRTEngineOp::Destroyer<T>>;
+TRTEngineOp::EngineCtxPair TRTEngineOp::get_engine(int batch_size,
+                                                   OpKernelContext* ctx,
+                                                   bool ignore_dim_change) {
```

Comment:
ignore_dim_change is not used?

Reply:
Engine generation logic needs expansion which is left to another PR due to already large size of this PR, where this will be needed

Reply:
Then please add a TODO.